File: usingJSON/src/notify.cpp
/*
 * notify.cpp
 *
 *  Created on: 20 Jun 2019
 *      Author: andrewh
 */
#include <stdio.h>
#include <stdlib.h>
#include <strings.h>
#include <mysql/mysql.h>

#include <sys/types.h>
#include <pwd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <mqueue.h>
#include <poll.h>
#include <time.h>

#include <iostream>
#include <fstream>
#include <sstream>

#include <string>
#include <vector>
#include <thread>
#include <set>

#include <nlohmann/json.hpp>
using json = nlohmann::json;

using namespace std;

map<string,string> getToMap(MYSQL *conn, string sqlCmd) {
	map<string,string> data;

	MYSQL_FIELD *field;

//	cout << sqlCmd << endl;

	int rc = mysql_query(conn, sqlCmd.c_str());

	MYSQL_RES *result = mysql_store_result( conn );
	MYSQL_ROW row = mysql_fetch_row(result);
	unsigned int num_fields = mysql_num_fields(result);
	unsigned int num_rows   = mysql_num_rows(result);

	if( num_rows > 0 ) {
		char *headers[num_fields];
		for(unsigned int i = 0; (field = mysql_fetch_field(result)); i++) {
//			cout << field->name ;
//			cout << ":" << row[i] << endl;

			data[ field->name ] = row[i];
		}
	}
	return data;
}

struct timestamp {
	long old;
	long stamp;
};
int main() {
	MYSQL *conn;
	map<string, string>result;

	map<string, struct timestamp>cache;
	bool firstTime=true;

	string cfgFile = "/etc/mqtt/bridge.json";

	ifstream cfgStream( cfgFile );
	json config = json::parse(cfgStream);

	string dbName = config["database"]["name"];
	string db     = config["database"]["db"];
	string user   = config["database"]["user"];
	string passwd = config["database"]["passwd"];

	conn=mysql_init(NULL);

	/* Connect to database */
	if (!mysql_real_connect(conn, dbName.c_str(), user.c_str(), passwd.c_str(), db.c_str(), 0, NULL, 0)) {
		fprintf(stderr, "%s\n", mysql_error(conn));
		return 1;
	}


	cache["mqtt"].old = 0;
	cache["mqtt"].stamp = 0;

	cache["internal"].old = 0;
	cache["internal"].stamp = 0;

	string sqlCmd;

	//	long now = (unsigned)time(NULL);
	long old=0;
	set<string> changed;
	long stamp;

	while(true) {
		for (pair<string, struct timestamp> element : cache) {
			string word = element.first;
			// Accessing VALUE from element.
			long count = element.second.old;
			cout << word << " :: " << count << endl;

			sqlCmd = "select name,state,unix_timestamp(logtime) from " + word + " order by logtime desc limit 1;";
			cout << sqlCmd << endl;

			result = getToMap(conn, sqlCmd) ;

			stamp = atol((result["unix_timestamp(logtime)"]).c_str());
			cache[word].stamp = stamp;

			if( stamp > cache[word].old ) {
				cout << "====== Check " + word << endl;
				sqlCmd="select name,state,unix_timestamp(logtime) from mqtt where logtime > from_unixtime(" + to_string(cache[word].old) + ");";
				cout << sqlCmd << endl;
				cache[word].old = stamp;
			}

		}
		sleep(10);
	}
	return 0;
}

File: usingJSON/Python/notify.py
#!/usr/bin/env python3

import sys 
import os.path
import time
import getopt

from pushbullet import Pushbullet
import pymysql as sql 
import json

db = None

connected=False

def getToMap(conn, sqlCmd):
    print("getToMap")

def stateToLogic( state ) :

    if state in ["ON","TRUE"]:
        return "TRUE"

    if state in ["OFF","FALSE"]:
        return "FALSE"

def main():
    global db
    configFile="/etc/mqtt/bridge.json"
    database = 'localhost'

    old = 0
    verbose = False
    testing = False

    cache = {}

    try:
        opts,args = getopt.getopt(sys.argv[1:], "c:hvt", ["config=","help","verbose","test"])
        for o,a in opts:
            if o in ["-h","--help"]:
                usage()
                sys.exit()
            elif o in ["-c","--config"]:
                configFile = a
            elif o in ["-v","--verbose"]:
                verbose=True
            elif o in ["-t","--test"]:
                testing=True

    except getopt.GetoptError as err:
        print(err)
        usage()
        sys.exit(2)

    token=""
    tokenFile = open("token.txt","r")

    if tokenFile.mode =="r":
        token=(tokenFile.read()).rstrip("\n\r")

    tokenFile.close()

    if testing == True:
        print("Would have sent " + token )
    else:
        pb = Pushbullet(token)

    if os.path.exists(configFile):
        with open( configFile, 'r') as f:
            cfg = json.load(f)

        database = cfg['database']['name']
        useDb    = cfg['database']['db']
        user     = cfg['database']['user']
        passwd   = cfg['database']['passwd']
    else:
        print(configFile + " not found..")
        print(".. using defaults")

#    cache['mqtt']     = {'old':0, 'stamp':0}
#    cache['internal'] = {'old':0, 'stamp':0}
    cache['io_point']     = {'old':0, 'stamp':0}

    print("Database is " + database)
    while True:
        db = sql.connect(database, user,passwd, useDb)
        sqlCmd=""
        for table in cache.keys():
            cursor = db.cursor()
            sqlCmd = "select name,state,unix_timestamp(logtime),notify from " + table + " order by logtime desc limit 1;"
            cursor.execute( sqlCmd )
            data = cursor.fetchone()
    
            name    = data[0]
            state   = data[1]
            logtime = data[2]
            notify  = data[3]
    
            if verbose:
                print(sqlCmd)
                print("Name   :" + name)
                print("State  :" + state)
                print("Logtime:" , logtime )
    
            if cache[ table ]['old'] == 0:
#                print("First time")
                cache[ table ]['old'] = logtime
                cache[ table ]['stamp'] = logtime
            else:
#                print("Been here before")
#                print("logtime:",logtime )
#                print("Old    :" , cache[ table ]['old']) 

                if logtime > cache[ table ]['old'] :

                    if verbose:
                        print( name + " was set to " + state + " at ", logtime )
                    cache[ table ]['old'] = logtime
#                    push = pb.push_note("This is the title", "This is the body")
                    if notify == 'YES':
                        if testing:
                            print("Would have published " + name + ":" + state)
                        else:
                            push = pb.push_note(name, state)
                            print(push)
            
        cursor.close()
        db.close()
        time.sleep(10)


main()
